src/stan/variational/advi.hpp

```diff
@@ -150,18 +157,170 @@ namespace stan {
       }
 
       /**
-       * Runs stochastic gradient ascent with Adagrad.
+       * Heuristic grid search to adapt eta to the scale of the problem.
        *
-       * @param  variational    variational distribution
-       * @param  tol_rel_obj    relative tolerance parameter for convergence
-       * @param  max_iterations max number of iterations to run algorithm
+       * @param[in] variational initial variational distribution.
+       * @param adapt_iterations number of iterations to spend doing stochastic
+       * gradient ascent at each proposed eta value.
+       * @return adapted (tuned) value of eta via heuristic grid search
+       * @throw std::domain_error If either (a) the initial ELBO cannot be
+       * computed at the initial variational distribution, (b) all step-size
+       * proposals in eta_sequence fail.
        */
-      void robbins_monro_adagrad(Q& variational,
+      double adapt_eta(Q& variational, int adapt_iterations) const {
+        static const char* function = "stan::variational::advi::adapt_eta";
+
+        stan::math::check_positive(function,
+                                   "Number of adaptation iterations",
+                                   adapt_iterations);
+
+
+        if (print_stream_) {
+          *print_stream_ << "Begin eta adaptation." << std::endl;
+        }
+
+        // Sequence of eta values to try during adaptation
+        std::queue<double> eta_sequence;
```

Comment:
You mean adjust the queue dynamically?  If not, it should be a static
constant array or std::vector with an index into the current position.



> On Nov 20, 2015, at 10:49 AM, Alp Kucukelbir <notifications@github.com> wrote:
> 
> In src/stan/variational/advi.hpp:
> 
> >         */
> > -      void robbins_monro_adagrad(Q& variational,
> > +      double adapt_eta(Q& variational, int adapt_iterations) const {
> > +        static const char* function = "stan::variational::advi::adapt_eta";
> > +
> > +        stan::math::check_positive(function,
> > +                                   "Number of adaptation iterations",
> > +                                   adapt_iterations);
> > +
> > +
> > +        if (print_stream_) {
> > +          *print_stream_ << "Begin eta adaptation." << std::endl;
> > +        }
> > +
> > +        // Sequence of eta values to try during adaptation
> > +        std::queue<double> eta_sequence;
> 
> we've been playing around with the values and the number of eta values to test. the queue made sense. we may still want to play around with eta values down the road. so i think the queue still makes sense.
> 
> —
> Reply to this email directly or view it on GitHub.
>